fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -134,43 +144,66 @@ ACTOR Future<Void> getRate(UID myID, Reference<AsyncVar<ServerDBInfo>> db, int64
 	}
 }
 
+struct TransactionRateInfo {
+	double rate;
+	double limit;
+
+	TransactionRateInfo(double rate) : rate(rate), limit(0) {}
+
+	void reset(double elapsed) {
+		limit =
+		    std::min(0.0, limit) + std::min(rate * elapsed, SERVER_KNOBS->START_TRANSACTION_MAX_TRANSACTIONS_TO_START);
+	}
+
+	bool canStart(int64_t numAlreadyStarted) { return numAlreadyStarted < limit; }
+
+	void updateBudget(int64_t numStarted) { limit -= numStarted; }
+};
+
 ACTOR Future<Void> queueTransactionStartRequests(
-	std::priority_queue< std::pair<GetReadVersionRequest, int64_t>, std::vector< std::pair<GetReadVersionRequest, int64_t> > > *transactionQueue,
-	FutureStream<GetReadVersionRequest> readVersionRequests,
-	PromiseStream<Void> GRVTimer, double *lastGRVTime,
-	double *GRVBatchTime, FutureStream<double> replyTimes,
-	ProxyStats* stats) 
-{
+    std::priority_queue<std::pair<GetReadVersionRequest, int64_t>,
+                        std::vector<std::pair<GetReadVersionRequest, int64_t>>>* transactionQueue,
+    FutureStream<GetReadVersionRequest> readVersionRequests, PromiseStream<Void> GRVTimer, double* lastGRVTime,
+    double* GRVBatchTime, FutureStream<double> replyTimes, ProxyStats* stats, TransactionRateInfo* batchRateInfo) {
 	state int64_t counter = 0;
 	loop choose{
 		when(GetReadVersionRequest req = waitNext(readVersionRequests)) {
-			if (req.debugID.present())
+			if (req.debugID.present()) {
 				g_traceBatch.addEvent("TransactionDebug", req.debugID.get().first(), "MasterProxyServer.queueTransactionStartRequests.Before");
+			}
 
-			stats->txnStartIn += req.transactionCount;
-			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE)
+			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE) {
 				stats->txnSystemPriorityStartIn += req.transactionCount;
-			else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT)
+			} else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT) {
 				stats->txnDefaultPriorityStartIn += req.transactionCount;
-			else
+			} else {
+				// Return error for bath_priority GRV requests
+				if (batchRateInfo->limit <= 1) {
```

Comment:
```suggestion
				if (batchRateInfo->limit < 1) {
```

Reply:
Because of the current nature of our rate limiter, I think this might lead to some undesired behavior. For example, I think it would be possible that the overall rate of batch work is acceptable, but if it isn't evenly spaced you might reject requests unnecessarily.

I have some intent to change how the limiter works so that you probably could more reliably perform a test like this (see https://github.com/apple/foundationdb/issues/2206). I also have a Python based model of this code that I've been using to run some experiments on, and we could possibly incorporate this logic in as well to measure unneeded rejections.

Reply:
I think there may also be an issue here when the rate is low enough that the limit would never be allowed to reach 1, even when no transactions are being started. Looking at current knobs, I think the limit can never exceed a proxy's rate per second divided by 100, so if each proxy is allotted fewer than 100 transactions per second for batch priority, we won't be able to start any batch transactions.

This could be mitigated in the current code by rejecting transactions when the limit is <=0 rather than 1, but it may not be worth investing too much time into this without first discussing what the future limiter is going to look like.

Reply:
> Looking at current knobs, I think the limit can never exceed a proxy's rate per second divided by 100, 

Where is this knob?

Reply:
This is the line that caps the limit:

https://github.com/apple/foundationdb/blob/493d39be7a49c83ac266fe33c7fd8a0894a80156/fdbserver/MasterProxyServer.actor.cpp#L1114

Here's the value of the knob:

https://github.com/apple/foundationdb/blob/493d39be7a49c83ac266fe33c7fd8a0894a80156/fdbserver/Knobs.cpp#L274

Reply:
Got it. I will leave this part as is until we have a rough idea about the future limiter. Will add a TODO in the code.